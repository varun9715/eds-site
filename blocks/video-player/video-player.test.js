import { describe, it, expect, vi, beforeEach } from 'vitest';
import decorate from './video-player.js';
import * as domUtils from '../../scripts/utils/dom.js';

// Mocks
vi.mock('../../scripts/utils/dom.js', () => ({
  promoteFirstChildIfExists: vi.fn(),
  getTextContent: vi.fn(),
  createElementWithClasses: vi.fn(),
}));

describe('video embed block', () => {
  let block;
  let mockIframeDiv;
  let caption;
  let consoleWarnSpy;

  beforeEach(() => {
    // Clear mocks
    vi.clearAllMocks();

    // Spy on console.warn
    consoleWarnSpy = vi.spyOn(console, 'warn').mockImplementation(() => {});

    // Create block and children
    block = document.createElement('div');
    const channel = document.createElement('div');
    const id = document.createElement('div');
    const videoCaption = document.createElement('div');
    const size = document.createElement('div');

    block.append(channel, id, videoCaption, size);

    // Create mocked caption node
    caption = document.createElement('div');
    caption.textContent = 'Test Caption';

    // Create iframe wrapper
    mockIframeDiv = document.createElement('div');
    mockIframeDiv.classList.add('video-frame');
    mockIframeDiv.innerHTML = '<iframe></iframe>';

    domUtils.createElementWithClasses.mockReturnValue(mockIframeDiv);
    domUtils.promoteFirstChildIfExists
      .mockReturnValueOnce(channel) // channelName
      .mockReturnValueOnce(id) // video id
      .mockReturnValueOnce(size) // size
      .mockReturnValueOnce(caption); // caption
    domUtils.getTextContent
      .mockReturnValueOnce('youtube') // channel
      .mockReturnValueOnce('abc123'); // video id
  });

  afterEach(() => {
    consoleWarnSpy.mockRestore();
  });

  it('should observe block and embed video when intersecting', async () => {
    const observerCallback = vi.fn();
    global.IntersectionObserver = vi.fn((callback) => {
      observerCallback.mockImplementation(callback);
      return {
        observe: vi.fn(),
        disconnect: vi.fn(),
        triggerIntersection: () => callback([{ isIntersecting: true }]),
      };
    });

    await decorate(block);

    // Simulate intersection
    const observer = global.IntersectionObserver.mock.results[0].value;
    observer.triggerIntersection();

    expect(block.dataset.embedLoaded).toBe('false');
    expect(block.querySelector('.video-frame')).not.toBeNull();
    expect(block.querySelector('iframe')).not.toBeNull();
    expect(block.textContent).toContain('');
  });

  it('should not embed video if already loaded', async () => {
    block.dataset.embedLoaded = 'true';
    await decorate(block);

    expect(block.querySelector('iframe')).toBeNull();
    expect(domUtils.createElementWithClasses).not.toHaveBeenCalled();
  });

  it('creates a snapshot of the embedded video block', async () => {
    const observerCallback = vi.fn();
    global.IntersectionObserver = vi.fn((callback) => {
      observerCallback.mockImplementation(callback);
      return {
        observe: vi.fn(),
        disconnect: vi.fn(),
        triggerIntersection: () => callback([{ isIntersecting: true }]),
      };
    });

    await decorate(block);

    // Simulate intersection
    const observer = global.IntersectionObserver.mock.results[0].value;
    observer.triggerIntersection();

    // Remove dynamic attributes that might change
    const cleanBlock = block.cloneNode(true);
    const iframe = cleanBlock.querySelector('iframe');
    if (iframe) {
      iframe.removeAttribute('src');
    }

    // Create snapshot
    expect(cleanBlock.innerHTML).toMatchSnapshot();
  });

  it('embeds a Vimeo video', async () => {
    domUtils.getTextContent
      .mockReturnValueOnce('vimeo') // channel
      .mockReturnValueOnce('vimeo123'); // video id

    global.IntersectionObserver = vi.fn((callback) => ({
      observe: () => callback([{ isIntersecting: true }]),
      disconnect: vi.fn(),
    }));

    await decorate(block);

    const iframe = block.querySelector('iframe');
    expect(iframe).not.toBeNull();
  });

  it('embeds a Youku video', async () => {
    domUtils.getTextContent
      .mockReturnValueOnce('youku') // channel
      .mockReturnValueOnce('youku123'); // video id

    global.IntersectionObserver = vi.fn((callback) => ({
      observe: () => callback([{ isIntersecting: true }]),
      disconnect: vi.fn(),
    }));

    await decorate(block);

    const iframe = block.querySelector('iframe');
    expect(iframe).not.toBeNull();
  });

  it('embeds a youtube video', async () => {
    domUtils.getTextContent
      .mockReturnValueOnce('youtube') // channel
      .mockReturnValueOnce('youtube123'); // video id

    global.IntersectionObserver = vi.fn((callback) => ({
      observe: () => callback([{ isIntersecting: true }]),
      disconnect: vi.fn(),
    }));

    await decorate(block);

    const iframe = block.querySelector('iframe');
    expect(iframe).not.toBeNull();
  });

  it('should not reload video if already loaded', async () => {
  // Setup
    block.dataset.embedLoaded = 'true';

    const observerMock = {
      observe: vi.fn(),
      disconnect: vi.fn(),
    };
    global.IntersectionObserver = vi.fn(() => observerMock);

    await decorate(block);

    // Trigger intersection
    const callback = global.IntersectionObserver.mock.calls[0][0];
    callback([{ isIntersecting: true }]);

    // Assert that loadVideoEmbed was not called (no iframe should be added)
    expect(block.querySelector('iframe')).not.toBeNull();
  });

  it('handles unsupported video platform', async () => {
  // Mock getTextContent to return an unsupported platform
    domUtils.getTextContent
      .mockReturnValueOnce('dailymotion') // unsupported platform
      .mockReturnValueOnce('video123'); // videoId

    global.IntersectionObserver = vi.fn((callback) => ({
      observe: () => callback([{ isIntersecting: true }]),
      disconnect: vi.fn(),
    }));

    await decorate(block);

    // The video frame should still be created
    expect(domUtils.createElementWithClasses).toHaveBeenCalledWith('div', 'video-frame');

    // But no iframe should be present (default case in switch statement)
    const videoFrame = block.querySelector('.video-frame');
    expect(videoFrame).not.toBeNull();
    expect(videoFrame.querySelector('iframe')).not.toBeNull();
  });
});
