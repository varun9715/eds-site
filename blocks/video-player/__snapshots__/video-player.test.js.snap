// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`video embed block > creates a snapshot of the embedded video block 1`] = `"<div class="video-frame"><iframe></iframe><iframe src="https://www.youtube.com/embed/abc123?cc_load_policy=1" title="Content from youtube" allowfullscreen="" style="width: 100%; height: 100%; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; border-radius: inherit;" allow="fullscreen; encrypted-media; accelerometer; gyroscope; picture-in-picture"></iframe>null</div>"`;
